# Neural Network with Full-Connected Layer

## Prologue

The classification problem was solved using a kind of Neural Network, which contains just Full-Connected Layers (named like Dense too). We used differents configurations and parameters to obtain the best result. Moreover, We used MNIST dataset which is provided by Keras.


Fixed parameters:

    - batch size = 128
    - epochs = 12
    - loss function = categorical crossentropy
    - optimizer = stocastic gradient descedent


## Libraries

In [1]:
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras import backend as K

/home/fernanda/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


## Dataset: Loading and preparation

In [6]:
# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# let's print the shape before we reshape and normalize
print("X_train shape", x_train.shape)
print("y_train shape", y_train.shape)
print("X_test shape", x_test.shape)
print("y_test shape", y_test.shape)

# building the input vector from the 28x28 pixels
x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')

# normalizing the data to help with the training
x_train /= 255
x_test /= 255

# print the final input shape ready for training
print("Train matrix shape", x_train.shape)
print("Test matrix shape", x_test.shape)

# convert class vectors to binary class matrices
num_classes = 10
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

#print (y_test.shape)
#print ('input_shape:', input_shape)

batch_size = 128
epochs = 12

X_train shape (60000, 28, 28)
y_train shape (60000,)
X_test shape (10000, 28, 28)
y_test shape (10000,)
Train matrix shape (60000, 784)
Test matrix shape (10000, 784)


# 1 - Dense + ReLu

<img src="I1.png">

In [7]:
model = Sequential()
#model.add(Flatten())

# Input Layer
input_shape=784
# Hidden Layer
model.add(Dense(512, activation='relu', input_shape=(784,)))
#model.add(Dense(1000, activation='relu'))
# Output Layer
model.add(Dense(num_classes, activation='relu'))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 512)               401920    
_________________________________________________________________
dense_4 (Dense)              (None, 10)                5130      
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________________________________________________________________


In [9]:
#model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='sgd')

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))

score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 3s 44us/step - loss: 3.9605 - acc: 0.1022 - val_loss: 3.6637 - val_acc: 0.1010
Epoch 2/12
60000/60000 [==============================] - 2s 41us/step - loss: 3.5389 - acc: 0.1023 - val_loss: 3.5300 - val_acc: 0.1012
Epoch 3/12
60000/60000 [==============================] - 2s 41us/step - loss: 3.4087 - acc: 0.1063 - val_loss: 3.3889 - val_acc: 0.1111
Epoch 4/12
60000/60000 [==============================] - 2s 41us/step - loss: 3.2229 - acc: 0.1685 - val_loss: 3.1190 - val_acc: 0.2811
Epoch 5/12
60000/60000 [==============================] - 2s 40us/step - loss: 1.9691 - acc: 0.4330 - val_loss: 1.1703 - val_acc: 0.6023
Epoch 6/12
60000/60000 [==============================] - 2s 41us/step - loss: 1.4491 - acc: 0.3497 - val_loss: 1.0918 - val_acc: 0.5322
Epoch 7/12
60000/60000 [==============================] - 2s 41us/step - loss: 1.0499 - acc: 0.6184 - val_loss: 0.8490 - val_acc

# 2 - Dense + ReLu + Dense + Softmax

<img src="I1_2.png">

In [10]:
model2 = Sequential()

# Input Layer
input_shape=784
# Hidden Layer
model2.add(Dense(512, activation='relu', input_shape=(784,)))
# Output Layer
model2.add(Dense(num_classes, activation='softmax'))
           
model2.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_5 (Dense)              (None, 512)               401920    
_________________________________________________________________
dense_6 (Dense)              (None, 10)                5130      
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________________________________________________________________


In [11]:
#model2.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')
model2.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='sgd')

model2.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))

score = model2.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 3s 46us/step - loss: 1.0770 - acc: 0.7703 - val_loss: 0.5884 - val_acc: 0.8685
Epoch 2/12
60000/60000 [==============================] - 2s 42us/step - loss: 0.5166 - acc: 0.8726 - val_loss: 0.4296 - val_acc: 0.8922
Epoch 3/12
60000/60000 [==============================] - 3s 42us/step - loss: 0.4188 - acc: 0.8892 - val_loss: 0.3701 - val_acc: 0.9022
Epoch 4/12
60000/60000 [==============================] - 3s 42us/step - loss: 0.3734 - acc: 0.8985 - val_loss: 0.3381 - val_acc: 0.9081
Epoch 5/12
60000/60000 [==============================] - 3s 43us/step - loss: 0.3454 - acc: 0.9049 - val_loss: 0.3172 - val_acc: 0.9147
Epoch 6/12
60000/60000 [==============================] - 3s 42us/step - loss: 0.3252 - acc: 0.9098 - val_loss: 0.3011 - val_acc: 0.9180
Epoch 7/12
60000/60000 [==============================] - 3s 44us/step - loss: 0.3094 - acc: 0.9141 - val_loss: 0.2878 - val_acc

# Conclusion

We could notice that the accuracy increased from 0.89 to 0.9312 when using the softmax activation function. It happened because this function normalizes the vector into a probability distribution. In other words, this function normalize the outputs in the interval between 0 and 1.

# References

- https://nextjournal.com/gkoehler/digit-recognition-with-keras